sklearn/model_selection/_split.py

```diff
@@ -1472,6 +1472,11 @@ def _iter_indices(self, X, y, groups=None):
         y = check_array(y, ensure_2d=False, dtype=None)
         n_train, n_test = _validate_shuffle_split(n_samples, self.test_size,
                                                   self.train_size)
+
+        if y.ndim == 2:
+            # for multi-label y, map each distinct row to its string repr:
+            y = np.array([str(row) for row in y])
```

Comment:
@jnothman str is guaranteed to do the right thing. It even seems faster on large y, but it seems like it would take up more memory (i need to install memit...)

```

In [10]: Y.shape
Out[10]: (1000, 10000)

In [11]: Y[:3, :3]
Out[11]:
array([[False,  True, False],
       [False,  True,  True],
       [False,  True,  True]], dtype=bool)

In [12]: %timeit [str(y) for y in Y]
10 loops, best of 3: 178 ms per loop

In [13]: %timeit [hash(tuple(y)) for y in Y]
1 loop, best of 3: 216 ms per loop

In [14]: Y = Y.astype(np.int)

In [15]: %timeit [str(y) for y in Y]
10 loops, best of 3: 180 ms per loop

In [16]: %timeit [hash(tuple(y)) for y in Y]
1 loop, best of 3: 580 ms per loop
```

Reply:
@vene Are you sure `str(row)` is the right thing to do here? I am finding that for a row with many elements, `str(row)` ends up being a string representing just the first 3 and last 3 entries:
```'[0 0 0 ..., 0 0 0]'```
I have a sparse target matrix `y` that is very wide and has lots of zeros, so `y = np.array([str(row) for row in y])` converts `y` to a very small number of strings that look like the one above. I think this behavior is not what was intended, right?

Reply:
@cbrummitt you're correct, this seems to trigger for vectors with over 1000 elements... Thanks for the catch!

Unfortunately `repr` has the same issue. A quick fix seems to be `" ".join(x.astype('str')`, which is even faster than `str` on arrays smaller than 1000 elements. Another one would be to use `murmurhash3_32`. Would you like to submit a fix?

Reply:
@vene I can try submitting a fix. It may take me several days to get around to it, so feel free to beat me to it if you want :)

Reply:
@vene Just submitted a PR: https://github.com/scikit-learn/scikit-learn/pull/9922